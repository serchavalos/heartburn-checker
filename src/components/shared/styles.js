export const minWidth = "500px";
export const fontColorPrimary = "#3f6072";
export const fontColorSecondary = "#7b98a9";
export const green = "#6accba";
export const gray = "#e8ecf0";
