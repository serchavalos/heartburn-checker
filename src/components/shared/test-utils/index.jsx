import { MemoryRouter } from "react-router-dom";

export const TestContainer = MemoryRouter;
