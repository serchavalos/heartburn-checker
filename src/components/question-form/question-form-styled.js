import styled from "styled-components";

import { green, gray } from "../shared/styles";
import checkMark from "../../assets/ic-checkmark.svg";

const lineHeight = 23;

export const Question = styled.div`
  display: block;
  margin-top: 4rem;
`;

export const Options = styled.ul`
  display: flex;
  margin: 1rem auto;
  justify-content: space-evenly;
  list-style: none;
`;

export const Option = styled.li`
  display: block;
  line-height: ${lineHeight}px;
  color: ${({ selected }) => (selected ? "white" : green)};
  background-color: ${({ selected }) => (selected ? green : "white")};
  border: ${({ selected }) => (selected ? "none" : `1px solid ${gray}`)};
  font-weight: bold;
  padding: 0.6rem 1.5rem;
  min-width: 8.5rem;
  border-radius: 1.5rem;
  text-align: ${({ selected }) => (selected ? "left" : "center")};
  cursor: pointer;
  outline: none;
  position: relative;

  &::after {
    content: "";
    display: ${({ selected }) => (selected ? "block" : "none")};
    background: url(${checkMark}) center center no-repeat white;
    height: ${lineHeight}px;
    width: ${lineHeight}px;
    margin-left: 0.5rem;
    border-radius: 50%;
    top: 50%;
    position: absolute;
    right: ${lineHeight / 2}px;
    margin-top: -${lineHeight / 2}px;
  }
`;

export const Footer = styled.div`
  width: 100%;
`;
