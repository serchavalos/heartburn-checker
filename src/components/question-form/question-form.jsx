import React from "react";
import PropTypes from "prop-types";

import { Question, Options, Option, Footer } from "./question-form-styled.js";
import { Title } from "../shared/title/title";
import { Button } from "../shared/button/button";

import questionManager from "../../services/question-manager";

export class QuestionForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentScore: props.score,
      selectedAnswer: null
    };
  }

  onAnwserClicked(answerId, initialScore) {
    const currentScore = parseInt(this.props.score, 10) + initialScore;
    this.setState({
      selectedAnswer: answerId,
      currentScore: currentScore
    });
  }

  render() {
    const { id, text, answers } = this.props;
    const { selectedAnswer = null, currentScore } = this.state;
    const [type, nextStepId] = selectedAnswer
      ? questionManager.getNextStep(id, selectedAnswer, currentScore)
      : [null, null];

    const path = !selectedAnswer
      ? ""
      : type === "question"
      ? `/question/${nextStepId}/score/${currentScore}`
      : `/outcome/${nextStepId}`;

    return (
      <React.Fragment>
        <Question>
          <Title>{text}</Title>

          <Options>
            {answers.map(({ id, label, score }) => (
              <Option
                key={id}
                selected={selectedAnswer === id}
                onClick={() => this.onAnwserClicked(id, score)}
              >
                {label}
              </Option>
            ))}
          </Options>
        </Question>

        <Footer>
          <Button
            to={path}
            onClick={evt => !selectedAnswer && evt.preventDefault()}
          >
            Next
          </Button>
        </Footer>
      </React.Fragment>
    );
  }
}

QuestionForm.propTypes = {
  id: PropTypes.string,
  text: PropTypes.string,
  answers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      label: PropTypes.string,
      score: PropTypes.number
    })
  )
};
