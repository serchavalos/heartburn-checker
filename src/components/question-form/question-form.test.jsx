import React from "react";
import { render } from "react-testing-library";

import { TestContainer } from "../shared/test-utils";
import { QuestionForm } from "./question-form.jsx";

describe("QuestionForm", () => {
  it("should render propperly", () => {
    const { container, getByText } = render(
      <TestContainer>
        <QuestionForm
          id="is_heartburn_known"
          text="Is your heartburn previously known?"
          answers={[
            {
              id: "is_heartburn_known_yes",
              label: "Yes",
              score: 5
            },
            {
              id: "is_heartburn_known_no",
              label: "No",
              score: 0
            }
          ]}
        />
      </TestContainer>
    );
    expect(container).toMatchSnapshot();
    expect(getByText("Is your heartburn previously known?")).toBeDefined();
    expect(container.querySelectorAll("li").length).toBe(2);
  });
});
