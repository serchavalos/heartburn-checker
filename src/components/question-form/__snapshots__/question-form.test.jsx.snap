// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionForm should render propperly 1`] = `
<div>
  <div
    class="sc-bdVaJa kXMpWJ"
  >
    <h1
      class="sc-ifAKCX gxTlJM"
    >
      Is your heartburn previously known?
    </h1>
    <ul
      class="sc-bwzfXH gtAMZT"
    >
      <li
        class="sc-htpNat idJYXX"
      >
        Yes
      </li>
      <li
        class="sc-htpNat idJYXX"
      >
        No
      </li>
    </ul>
  </div>
  <div
    class="sc-bxivhb jcNPuI"
  >
    <a
      class="sc-EHOje cTWAwk"
      href="/"
    >
      Next
    </a>
  </div>
</div>
`;
